libr/util/charset.c

```diff
@@ -113,24 +113,36 @@ R_API size_t r_charset_encode_str(RCharset *rc, ut8 *out, size_t out_len, const
 
 // assumes out is as big as in_len
 R_API size_t r_charset_decode_str(RCharset *rc, ut8 *out, size_t out_len, const ut8 *in, size_t in_len) {
-	char k[32];
 	char *o = (char*)out;
-	int i;
-	for (i = 0; i < in_len; i++) {
-		ut8 ch_in = in[i];
-
-		//zero terminate the string
-		snprintf (k, sizeof (k), "%c", ch_in);//snprintf (k, sizeof (k), "0x%02x", ch_in);
-		char *v = sdb_get (rc->db_char_to_hex, k, 0);
-		memmove(v, v+2, strlen (v));
-
-		//convert to ascii
-		char str_hx[32];
-		snprintf (str_hx, sizeof (str_hx), "%c", (char) strtol( v, 0, 16));
-		const char *ret = r_str_get_fail (str_hx, "?");
-		strcpy (o, ret);
-
-		o += strlen (o);
+	int j;
+	ut8 str[320];
+	size_t last_char_size;
+
+	size_t cur;
+	for (cur = 0; cur < in_len; ) {
+		memcpy (str, in, sizeof (str));
```

Comment:
remove this line is wrong.causes overflow and its unnecssary

Reply:
wait i think i got your idea to find the nth chars in the database, this can be very expensive as it depends on the size of str. and 320 is still too much, i would say 8 or 16 for now.

Reply:
I do not think this is unnecessary. we just need to find a way to come back to the previous state and to not allocate too much memory.

Reply:
the previous state is not necessary, you just trim one char from the end of the string on each iteration. so better use
```
 char *str = r_str_ndup(in, maxkeylen);
for (j ...
free(str);
```

Reply:
I am going to try.